src/algebra/group/hom.lean

```diff
@@ -81,35 +133,195 @@ attribute [nolint doc_blame] add_monoid_hom.to_zero_hom
+When you extend this structure, make sure to extend `mul_hom_class`.
+-/
 @[to_additive]
 structure mul_hom (M : Type*) (N : Type*) [has_mul M] [has_mul N] :=
 (to_fun : M → N)
 (map_mul' : ∀ x y, to_fun (x * y) = to_fun x * to_fun y)
 
-/-- Bundled monoid homomorphisms; use this for bundled group homomorphisms too. -/
+/-- `mul_hom_class F M N` states that `F` is a type of multiplication-preserving homomorphisms.
+
+You should declare an instance of this typeclass when you extend `mul_hom`.
+-/
+@[to_additive]
+class mul_hom_class (F : Type*) (M N : out_param $ Type*)
+  [has_mul M] [has_mul N] extends fun_like F M (λ _, N) :=
+(map_mul : ∀ (f : F) (x y : M), f (x * y) = f x * f y)
+
+@[to_additive]
+instance mul_hom.mul_hom_class : mul_hom_class (mul_hom M N) M N :=
+{ coe := mul_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_mul := mul_hom.map_mul' }
+
+@[simp, to_additive] lemma map_mul [mul_hom_class F M N] (f : F) (x y : M) :
+  f (x * y) = f x * f y :=
+mul_hom_class.map_mul f x y
+
+end mul
+
+section mul_one
+
+variables [mul_one_class M] [mul_one_class N]
+
+/-- `M →* N` is the type of functions `M → N` that preserve the `monoid` structure.
+`monoid_hom` is also used for group homomorphisms.
+
+When possible, instead of parametrizing results over `(f : M →+ N)`,
+you should parametrize over `(F : Type*) [monoid_hom_class F M N] (f : F)`.
+
+When you extend this structure, make sure to extend `monoid_hom_class`.
+-/
 @[ancestor one_hom mul_hom, to_additive]
 structure monoid_hom (M : Type*) (N : Type*) [mul_one_class M] [mul_one_class N]
   extends one_hom M N, mul_hom M N
 
-/-- Bundled monoid with zero homomorphisms; use this for bundled group with zero homomorphisms
-too. -/
+attribute [nolint doc_blame] monoid_hom.to_mul_hom
+attribute [nolint doc_blame] monoid_hom.to_one_hom
+
+infixr ` →* `:25 := monoid_hom
+
+/-- `monoid_hom_class F M N` states that `F` is a type of `monoid`-preserving homomorphisms.
+You should also extend this typeclass when you extend `monoid_hom`.
+-/
+@[ancestor mul_hom_class one_hom_class, to_additive]
+class monoid_hom_class (F : Type*) (M N : out_param $ Type*)
+  [mul_one_class M] [mul_one_class N]
+  extends mul_hom_class F M N, one_hom_class F M N
+
+@[to_additive]
+instance monoid_hom.monoid_hom_class : monoid_hom_class (M →* N) M N :=
+{ coe := monoid_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_mul := monoid_hom.map_mul',
+  map_one := monoid_hom.map_one' }
+
+@[to_additive]
+instance [monoid_hom_class F M N] : has_coe_t F (M →* N) :=
+⟨λ f, { to_fun := f, map_one' := map_one f, map_mul' := map_mul f }⟩
+
+@[to_additive]
+lemma map_mul_eq_one [monoid_hom_class F M N] (f : F) {a b : M} (h : a * b = 1) :
+  f a * f b = 1 :=
+by rw [← map_mul, h, map_one]
+
+/-- Group homomorphisms preserve inverse. -/
+@[simp, to_additive]
+theorem map_inv [group G] [group H] [monoid_hom_class F G H]
```

Comment:
Should this now in fact be proven for `mul_hom_class`? Because that's all you need.